simphony/cuds/model.py

```diff
@@ -39,119 +55,179 @@ def __init__(self):
-            if component_id is None:
-                component_id = uuid.uuid4()
+        # Only accept CUDSComponent subclasses and datasets
+        if not (isinstance(component, api.CUDSComponent) or
+                self._is_dataset(component)):
+            raise TypeError('Not a CUDSComponent nor a dataset object: %s.' %
+                            type(component))
+
+        # Do not accept items with duplicate names.
+        # Components/datasets with no name will be added, however
+        # it is not possible to get/remove them with `name`
+        if component.name in self._name_to_id_map:
+            raise ValueError('Name clash. A %s component has the same name %s'
+                             % (type(self.get(component.name)),
+                                component.name))
 
         # Add datasets separately
         if self._is_dataset(component):
+            if not component.name:
+                raise TypeError('Dataset must have a name.')
+
             self._dataset_store.add(component)
-            # Datasets have a uuid field called uid
+            # Datasets at the moment do not have uid
             self._map[component.name] = \
                 lambda key=component.name: self._dataset_store.get(key)
+            # Datasets use name as id
+            self._name_to_id_map[component.name] = component.name
         else:
-            # Store the component. Any cuds item has uuid property.
-            self._store[component_id] = component
-            self._map[component_id] = \
-                lambda key=component_id: self._store.get(key)
+            # Delete existing item with the same name
+            # Store the component. Any CUDS item has uid property.
+            self._store[component.uid] = component
+            self._map[component.uid] = \
+                lambda key=component.uid: self._store.get(key)
+            # Store name for name-to-id mapping.
+            # Components with no name will are not added to the mapping
+            # and it is not possible to access them using add/remove
+            # methods. Use `get_by_uid` and `remove_by_uid` instead.
+            if component.name not in (None, ''):
+                self._name_to_id_map[component.name] = component.uid
+
+    def get(self, name):
+        """Get the corresponding component from the CUDS computational model.
+
+        Parameters
+        ----------
+        name: str
+            name of the component
 
-        # If the component already has a defined uid, this just reassigns
-        # the same value.  If the component.uuid is originaly None, this
-        # assigns the new uid to the component.uuid
-        # Only do so after successfully adding the component
-        if hasattr(component, "uuid"):
-            component.uuid = component_id
+        Returns
+        -------
+        component: CUDSComponent
+            the corresponding CUDS component or None
+        """
+        uid = self._name_to_id_map.get(name)
```

Comment:
This was indeed on purpose. Trying to `get` a nonexistent object yields nothing, like python dict. User can always check the returned value, I think it is more elegant compared to catching an exception. However, since `remove` is not supposed to return anything, an error is desirable in case of a nonexistent key.